src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/UseSpanBasedStringConcat.cs

```diff
@@ -0,0 +1,310 @@
+            RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        private const string SubstringName = nameof(string.Substring);
+        private const string AsSpanName = nameof(MemoryExtensions.AsSpan);
+        private const string ConcatName = nameof(string.Concat);
+
+        /// <summary>
+        /// If the specified binary operation is a string concatenation operation, we try to walk up to the top-most
+        /// string-concatenation operation that it is part of. If it is not a string-concatenation operation, we simply
+        /// return false.
+        /// </summary>
+        private protected abstract bool TryGetTopMostConcatOperation(IBinaryOperation binaryOperation, [NotNullWhen(true)] out IBinaryOperation? rootBinaryOperation);
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!RequiredSymbols.TryGetSymbols(context.Compilation, out RequiredSymbols symbols))
+                return;
+
+            context.RegisterOperationBlockStartAction(OnOperationBlockStart);
+
+            void OnOperationBlockStart(OperationBlockStartAnalysisContext context)
+            {
+                //  Maintain set of all top-most concat operations so we don't report sub-expressions of an
+                //  already-reported violation. 
+                //  We also don't report any diagnostic if the concat operation has too many operands for the span-based
+                //  Concat overloads to handle.
+                var topMostConcatOperations = PooledConcurrentSet<IBinaryOperation>.GetInstance();
+
+                context.RegisterOperationAction(PopulateTopMostConcatOperations, OperationKind.Binary);
+                context.RegisterOperationBlockEndAction(ReportDiagnosticsOnRootConcatOperationsWithSubstringCalls);
+
+                void PopulateTopMostConcatOperations(OperationAnalysisContext context)
+                {
+                    //  If the current operation is a string-concatenation operation, walk up to the top-most concat
+                    //  operation and add it to the set.
+                    var binary = (IBinaryOperation)context.Operation;
+                    if (!TryGetTopMostConcatOperation(binary, out var topMostConcatOperation))
```

Comment:
This approach seems fine, but seems you can do without `topMostConcatOperations` and reporting in the end action. Instead you can just have the following in `IBinaryOperation` callback:

```csharp
  if (IsTopMostConcatOperation(binary) && ShouldBeReported(binary))
  {
      context.ReportDiagnostic(binary.CreateDiagnostic(Rule));
  }
```

Reply:
The problem with this approach is that we may end up reporting the same violation multiple times. I remember encountering this issue while writing the analyzer.